File: 2026/09-September/18-Friday/solution.py
class Solution:
    def maxNumOfSubstrings(self, s: str):
        n = len(s)

        # First and last occurrence of every character
        first = [n] * 26
        last = [-1] * 26

        for i, ch in enumerate(s):
            idx = ord(ch) - ord('a')
            first[idx] = min(first[idx], i)
            last[idx] = i

        # Find the smallest valid interval starting from position i
        intervals = []

        for i in range(n):
            idx = ord(s[i]) - ord('a')

            # Only start from the first occurrence of a character
            if i != first[idx]:
                continue

            l, r = i, last[idx]
            j = l
            valid = True

            while j <= r:
                c = ord(s[j]) - ord('a')

                # This character appeared before l,
                # so the substring cannot be valid.
                if first[c] < l:
                    valid = False
                    break

                r = max(r, last[c])
                j += 1

            if valid:
                intervals.append((l, r))

        # Choose non-overlapping intervals with earliest ending position
        intervals.sort(key=lambda x: x[1])

        ans = []
        end = -1

        for l, r in intervals:
            if l > end:
                ans.append(s[l:r + 1])
                end = r

        return ans

File: 2026/09-September/18-Friday/solution.java
import java.util.*;

class Solution {
    public List<String> maxNumOfSubstrings(String s) {
        int n = s.length();

        // First and last occurrence of each character
        int[] first = new int[26];
        int[] last = new int[26];

        Arrays.fill(first, n);
        Arrays.fill(last, -1);

        for (int i = 0; i < n; i++) {
            int idx = s.charAt(i) - 'a';
            first[idx] = Math.min(first[idx], i);
            last[idx] = i;
        }

        List<int[]> intervals = new ArrayList<>();

        // Find all valid intervals
        for (int i = 0; i < n; i++) {
            int idx = s.charAt(i) - 'a';

            // Only start from the first occurrence
            if (i != first[idx]) {
                continue;
            }

            int l = i;
            int r = last[idx];
            boolean valid = true;

            for (int j = l; j <= r; j++) {
                int c = s.charAt(j) - 'a';

                // Character appears before l
                if (first[c] < l) {
                    valid = false;
                    break;
                }

                // Expand interval
                r = Math.max(r, last[c]);
            }

            if (valid) {
                intervals.add(new int[]{l, r});
            }
        }

        // Sort by ending position
        intervals.sort((a, b) -> Integer.compare(a[1], b[1]));

        List<String> ans = new ArrayList<>();
        int end = -1;

        // Greedily choose non-overlapping intervals
        for (int[] interval : intervals) {
            int l = interval[0];
            int r = interval[1];

            if (l > end) {
                ans.add(s.substring(l, r + 1));
                end = r;
            }
        }

        return ans;
    }
}
